### PR TITLE
AAE-22465 Added logic to validate the cardinality

### PR DESCRIPTION
**Pull request by AkhileshPamidimarthi**
**What kind of change does this PR introduce?**

> - [x] Bugfix
> - [ ] Feature
> - [ ] Code style update (formatting, local variables)
> - [ ] Refactoring (no functional changes, no api changes)
> - [ ] Build related changes
> - [ ] Documentation
> - [ ] Other... Please describe:

## Description
 Added logic to validate the cardinality

- Issue Link: https://hyland.atlassian.net/browse/AAE-22465

**Does this PR introduce a breaking change?** (check one with "x")

> - [ ] Yes
> - [x] No

### CODE REVIEW

**Review by eromano — CHANGES_REQUESTED** [review]
I would prefer review our approach here with @gicappa @igdianov @erdemedeiros, let's have a call next week

File: activiti-core/activiti-process-validation/src/main/java/org/activiti/validation/validator/Problems.java [hunk 1]
```diff
@@ -100,6 +100,7 @@ public interface Problems {
   String THROW_EVENT_INVALID_EVENTDEFINITION = "activiti-throw-event-invalid-eventdefinition";
 
   String MULTI_INSTANCE_MISSING_COLLECTION = "activiti-multi-instance-missing-collection";
+    String MULTI_INSTANCE_HIGHER_CARDINALITY = "activiti-multi-instance-cardinality-greater-than-200";
```
Comment (jesty) [inline activiti-core/activiti-process-validation/src/main/java/org/activiti/validation/validator/Problems.java]: can u fix formatting? Thanks!

File: activiti-core/activiti-process-validation/src/main/java/org/activiti/validation/validator/impl/FlowElementValidator.java [hunk 2]
```diff
@@ -71,6 +71,19 @@ protected void handleMultiInstanceLoopCharacteristics(Process process, Activity
 			  addError(errors, Problems.MULTI_INSTANCE_MISSING_COLLECTION, process, activity,
 	    			"Either loopCardinality or loopDataInputRef/activiti:collection must been set");
 	    }
+            if(!StringUtils.isEmpty(multiInstanceLoopCharacteristics.getLoopCardinality())){
+                try {
+                    int loopCardinality=  Integer.parseInt(multiInstanceLoopCharacteristics.getLoopCardinality());
+                    if(loopCardinality >200){
+                        addError(errors, Problems.MULTI_INSTANCE_HIGHER_CARDINALITY, process, activity,
```
Comment (jesty) [inline activiti-core/activiti-process-validation/src/main/java/org/activiti/validation/validator/impl/FlowElementValidator.java]: please replace addError with addWarning

**Review by jesty — CHANGES_REQUESTED** [review]

File: activiti-core/activiti-process-validation/src/main/java/org/activiti/validation/validator/impl/FlowElementValidator.java [hunk 3]
```diff
@@ -37,61 +37,74 @@
...
+            if (StringUtils.isEmpty(multiInstanceLoopCharacteristics.getLoopCardinality())
+                && StringUtils.isEmpty(multiInstanceLoopCharacteristics.getInputDataItem())) {
+
+                addError(errors, Problems.MULTI_INSTANCE_MISSING_COLLECTION, process, activity,
+                    "Either loopCardinality or loopDataInputRef/activiti:collection must been set");
+            }
+            if(!StringUtils.isEmpty(multiInstanceLoopCharacteristics.getLoopCardinality())){
+                try {
+                    int loopCardinality=  Integer.parseInt(multiInstanceLoopCharacteristics.getLoopCardinality());
+                    if(loopCardinality >200){
```
Comment (jesty) [inline activiti-core/activiti-process-validation/src/main/java/org/activiti/validation/validator/impl/FlowElementValidator.java]: can we extract the value in an external property? thanks

**Review by jesty — CHANGES_REQUESTED** [review]

File: activiti-core/activiti-process-validation/src/main/java/org/activiti/validation/validator/impl/FlowElementValidator.java [hunk 4]
```diff
@@ -37,61 +39,76 @@
...
-	  if (activity.getDataOutputAssociations() != null) {
-	  	for (DataAssociation dataAssociation : activity.getDataOutputAssociations()) {
-	  		if (StringUtils.isEmpty(dataAssociation.getTargetRef())) {
-	  			 addError(errors, Problems.DATA_ASSOCIATION_MISSING_TARGETREF, process, activity,
-	  					 "Targetref is required on a data association");
-	  	    }
-	  	}
-	  }
+    protected static final int ID_MAX_LENGTH = 255;
+    @Value("${loop.cardinality.threshold}")
```
Comment (jesty) [inline activiti-core/activiti-process-validation/src/main/java/org/activiti/validation/validator/impl/FlowElementValidator.java]: add a blank new line above, prefix it with acitivi and specify the default value here.
```suggestion

    @Value("${activiti.loop.cardinality.threshold:200}")
```

File: activiti-core/activiti-process-validation/src/main/resources/application.properties [hunk 5]
```diff
@@ -0,0 +1 @@
+loop.cardinality.threshold=200
```
Comment (jesty) [inline activiti-core/activiti-process-validation/src/main/resources/application.properties]: specifying the default value, then, we can remove it

**Review by jesty — CHANGES_REQUESTED** [review]

**Comment by jesty** [comment]
> I would prefer review our approach here with @gicappa @igdianov @erdemedeiros, let's have a call next week

Hi @eromano we changed to warning. @gicappa @igdianov @erdemedeiros what do you think? 

I asked to add a warning because in my tests I noticed that when we have many parallelel task to create we can face 500 errors because the operation it's going to be really long.

File: activiti-core/activiti-process-validation/pom.xml [hunk 6]
```diff
@@ -37,5 +37,9 @@
       <artifactId>assertj-core</artifactId>
       <scope>test</scope>
     </dependency>
+      <dependency>
```
Comment (igdianov) [inline activiti-core/activiti-process-validation/pom.xml:40]: Is it really needed to include the dependency from Spring in the core engine?

File: activiti-core/activiti-process-validation/src/main/java/org/activiti/validation/validator/impl/FlowElementValidator.java [hunk 7]
```diff
@@ -37,61 +39,77 @@
...
-	  	for (DataAssociation dataAssociation : activity.getDataOutputAssociations()) {
-	  		if (StringUtils.isEmpty(dataAssociation.getTargetRef())) {
-	  			 addError(errors, Problems.DATA_ASSOCIATION_MISSING_TARGETREF, process, activity,
-	  					 "Targetref is required on a data association");
-	  	    }
-	  	}
-	  }
+    protected static final int ID_MAX_LENGTH = 255;
+
+    @Value("${loop.cardinality.threshold:200}")
```
Comment (igdianov) [inline activiti-core/activiti-process-validation/src/main/java/org/activiti/validation/validator/impl/FlowElementValidator.java:44]: It is not a good practice to use Spring value injection. The core validation configuration approach should not rely on Spring field injection to configure parameter limits.

File: activiti-core/activiti-process-validation/src/main/java/org/activiti/validation/validator/impl/FlowElementValidator.java [hunk 8]
```diff
@@ -37,61 +39,77 @@
...
-	  	}
-	  }
-	  if (activity.getDataOutputAssociations() != null) {
-	  	for (DataAssociation dataAssociation : activity.getDataOutputAssociations()) {
-	  		if (StringUtils.isEmpty(dataAssociation.getTargetRef())) {
-	  			 addError(errors, Problems.DATA_ASSOCIATION_MISSING_TARGETREF, process, activity,
-	  					 "Targetref is required on a data association");
-	  	    }
-	  	}
-	  }
```
Comment (igdianov) [inline activiti-core/activiti-process-validation/src/main/java/org/activiti/validation/validator/impl/FlowElementValidator.java:94]: Could you restore original indents to keep the formatting consistent for easier code review?

**Review by igdianov — CHANGES_REQUESTED** [review]
@jesty Could add unit tests to make sure there is coverage against future regressions?

**Comment by jesty** [comment]
We are going to put the async flag in the UI

**Pull request closed without merging**